onnx/defs/math/defs.cc

```diff
@@ -1665,4 +1666,130 @@ ONNX_OPERATOR_SET_SCHEMA(
+  auto mid_index = equation.find("->");
+  if (mid_index != std::string::npos) {
+    // Separate right and left hand sides of the equation
+    left_equation = equation.substr(0, mid_index);
+  } else {
+    // No right hand side
+    left_equation = equation;
+  }
+
+  std::string term;
+  size_t num_operands = 0;
+  size_t num_ellipsis = 0;
+  size_t num_ellipsis_indices = 0;
+
+  // Parse the left-hand side
+  std::stringstream str(left_equation);
+  while(std::getline(str, term, ',')) {
+    auto ellipsis_index = term.find("...");
+    if (ellipsis_index != std::string::npos) {
+      // If there is an ellipsis, the number of dimensions it represents must be total dim - letter dimensions
+      size_t rank = ctx.getInputType(num_operands)->tensor_type().shape().dim_size();
+      if (num_ellipsis == 0) {
+        num_ellipsis_indices = rank - term.size() + 3;
+      } else { // ellipsis has been seen before. Check that if dimensions are compatible
+        if (num_ellipsis_indices != rank - term.size() + 3) {
+          fail_shape_inference("Ellipsis represents incompatible dimensions.");
+        }
+      }
+      num_ellipsis++;
+    }
+    num_operands++;
+  }
+
+  if (numInputs != num_operands) {
+    fail_shape_inference("Number of input tensors does not match the operand in the equation.");
+  }
+
+  const size_t number_of_letters = 26;
+  size_t num_letter_occurrences[number_of_letters] = {0};
+  // Parse the provided right-hand side
+  if (mid_index != std::string::npos) {
+    std::string right_equation = equation.substr(mid_index + 2);
+    auto right_ellipsis_index = right_equation.find("...");
+    if (right_ellipsis_index != std::string::npos) { // Right-hand side contains ellipsis
+      for (size_t i = 0; i < num_ellipsis; ++i) {
+        output_shape->add_dim();
+      }
+    }
+    for (char c: right_equation) { // Add a dimension per each character in right hand equation
+      if (c != '.') {
+        output_shape->add_dim();
+      }
+    }
+  } else { // Infer the dimension for right-hand side
+    // If there's an ellipsis, add it's corresponding dimensions
+    for (size_t i = 0; i < num_ellipsis_indices; i++) {
+      output_shape->add_dim();
+    }
+    for (size_t i = 0; i < left_equation.size(); i++) { // Count chars that appear exactly once on left hand side
+      if ((left_equation.at(i) != ',') && (left_equation.at(i) != '.')) {
+        num_letter_occurrences[left_equation.at(i) - 'a']++;
+      }
+    }
+    for (size_t index = 0; index < number_of_letters; index++) {
+      if (num_letter_occurrences[index] == 1) {
+        output_shape->add_dim();
+      }
+    }
+  }
+}
+
+static const char* Einsum_ver12_doc = R"DOC(
+The Einsum operator evaluates algebraic tensor operations on a sequence of tensors, using the Einstein summation
+convention. The equation string contains a comma-separated sequence of lower case letters. Each term corresponds to
```

Comment:
Adding an explanation as below (please correct it as appropriate) right at the beginning may be helpful:

An einsum of the form `term1, term2, term3 -> output-term` produces an output tensor using the following equation:
```
   output[output-term] = implicit-sum( input1[term1] * input2[term] * input3[term3])
```
where the implicit-sum performs a summation over all the indices occurring in in the input terms (term1, term2, and term3) that do not occur in the output-term.

Reply:
Addressed this.

Reply:
Thanks. I still think that it is better to say "output[output-term]" instead of "output" in the equation.